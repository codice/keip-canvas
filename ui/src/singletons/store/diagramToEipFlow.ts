import isDeepEqual from "fast-deep-equal"
import { useStoreWithEqualityFn } from "zustand/traditional"
import {
  CustomEdge,
  CustomNode,
  CustomNodeType,
  DynamicEdge,
  EipFlowNode,
  isDynamicEdge,
  isFollowerNode,
  RouterKey,
} from "../../api/flow"
import {
  Attributes,
  EipChildNode,
  EipFlow,
  EipId,
  EipNode,
  FlowEdge,
} from "../../api/generated/eipFlow"
import {
  CHANNEL_ATTR_NAME,
  DEFAULT_OUTPUT_CHANNEL_NAME,
  lookupContentBasedRouterKeys,
  lookupEipComponent,
} from "../eipDefinitions"
import { describeFollower, describeFollowerFromId } from "../followerNodeDefs"
import { AppStore, EipConfig } from "./api"
import { useAppStore } from "./appStore"
import { childrenBreadthTraversal, getEipId } from "./storeViews"

const EIP_NAMESPACE_TO_XML_PREFIX: Record<string, string> = {
  xml: "int-xml",
  "web-services": "ws",
}

export const useEipFlow = () =>
  useStoreWithEqualityFn(
    useAppStore,
    (state) => diagramToEipFlow(state),
    isDeepEqual
  )

// TODO: Dynamic router logic is over-complicating this method, extract to
// a separate method and apply modifications after building original flow.
const diagramToEipFlow = (state: AppStore): EipFlow => {
  const nodeLookup = createNodeLookupMap(state.nodes)

  const routerChildMap = new Map<string, EipChildNode[]>()
  const routerAttrMap = new Map<string, Attributes>()

  const allEdges = addHiddenFollowerEdges(state.nodes, state.edges)

  const edges: FlowEdge[] = allEdges.map((edge) => {
    const source = getNodeId(edge.source, nodeLookup)
    const target = getNodeId(edge.target, nodeLookup)
    const edgeId = `ch-${source}-${target}`

    if (isDynamicEdge(edge)) {
      addRouterChannelMapping(edgeId, edge, routerChildMap, routerAttrMap)
    }

    return {
      id: edgeId,
      source,
      target,
      type: edge.sourceHandle === "discard" ? "discard" : "default",
    }
  })

  const nodes: EipNode[] = state.nodes.map((node) => {
    const eipId = getEipId(node.id)!
    const eipComponent = lookupEipComponent(eipId)!
    const namespace = eipId.namespace

    const routerKey = state.eipConfigs[node.id].routerKey
    const routerKeyAttrs =
      eipComponent.connectionType === "content_based_router" && routerKey
        ? getRouterKeyAttributes(eipId, routerKey)
        : {}

    return {
      id: getNodeId(node.id, nodeLookup),
      eipId: {
        ...eipId,
        namespace: EIP_NAMESPACE_TO_XML_PREFIX[namespace] ?? namespace,
      },
      description: state.eipConfigs[node.id]?.description,
      role: eipComponent.role,
      connectionType: eipComponent.connectionType,
      attributes: {
        ...state.eipConfigs[node.id]?.attributes,
        ...routerKeyAttrs.attributes,
        ...routerAttrMap.get(node.id),
      },
      children: buildChildTree(node.id, state.eipConfigs).concat(
        routerKeyAttrs.child ?? [],
        routerChildMap.get(node.id) ?? []
      ),
    }
  })

  return { nodes, edges }
}

const getNodeId = (nodeId: string, nodeLookup: Map<string, CustomNode>) => {
  const node = nodeLookup.get(nodeId)
  if (!node) {
    console.error(`Could not find node with id: ${nodeId}`)
    return nodeId
  }

  switch (node.type) {
    case CustomNodeType.EipNode: {
      return node.data.label ? node.data.label : node.id
    }
    case CustomNodeType.FollowerNode: {
      const leaderEipId = getEipId(node.data.leaderId)
      const descriptor = leaderEipId && describeFollower(leaderEipId)
      if (!descriptor) {
        return node.id
      }
      const leaderNode = nodeLookup.get(node.data.leaderId) as EipFlowNode
      return descriptor.generateLabel(
        leaderNode?.data.label ?? node.data.leaderId
      )
    }
  }
}

const addHiddenFollowerEdges = (nodes: CustomNode[], edges: CustomEdge[]) => {
  const combinedEdges = [...edges]
  nodes.forEach((node) => {
    if (isFollowerNode(node)) {
      const followerDesc = describeFollowerFromId(node.data.leaderId)
      if (followerDesc?.hiddenEdges) {
        const edges = followerDesc.hiddenEdges(
          node.data.leaderId,
          node.id
        ) as CustomEdge[]
        combinedEdges.push(...edges)
      }
    }
  })

  return combinedEdges
}

const buildChildTree = (
  rootId: string,
  eipConfigs: AppStore["eipConfigs"]
): EipChildNode[] => {
  const nodes: Record<string, EipChildNode> = {}

  for (const child of childrenBreadthTraversal(rootId)) {
    const currNode = childConfigToNode(eipConfigs[child.id])
    nodes[child.id] = currNode

    if (child.parentId !== null) {
      const parentNode = nodes[child.parentId]
      parentNode.children
        ? parentNode.children.push(currNode)
        : (parentNode.children = [currNode])
    }
  }

  return nodes[rootId].children ?? []
}

const childConfigToNode = (config: EipConfig): EipChildNode => ({
  name: config.eipId.name,
  attributes: config.attributes,
})

const createNodeLookupMap = (nodes: CustomNode[]) => {
  const map = new Map<string, CustomNode>()
  nodes.forEach((node) => map.set(node.id, node))
  return map
}

// TODO: consider extracting the ChannelMapping and RouterKey logic to the XML translation backend
const addRouterChannelMapping = (
  channelId: string,
  edge: DynamicEdge,
  routerChildMap: Map<string, EipChildNode[]>,
  routerAttributeMap: Map<string, Attributes>
) => {
  const mapping = edge.data?.mapping
  if (!mapping) {
    return
  }

  if (mapping.isDefaultMapping) {
    routerAttributeMap.set(edge.source, {
      [DEFAULT_OUTPUT_CHANNEL_NAME]: channelId,
    })
  } else {
    const child: EipChildNode = {
      name: mapping.mapperName,
      attributes: {
        [CHANNEL_ATTR_NAME]: channelId,
      },
    }
    if (mapping.matcherValue) {
      child.attributes![mapping.matcher.name] = mapping.matcherValue
    }
    const curr = routerChildMap.get(edge.source) ?? []
    routerChildMap.set(edge.source, curr.concat(child))
  }
}

const getRouterKeyAttributes = (eipId: EipId, routerKey: RouterKey) => {
  const routerKeyDef = lookupContentBasedRouterKeys(eipId)
  if (!routerKeyDef) {
    return {}
  }
  switch (routerKeyDef.type) {
    case "attribute": {
      return { attributes: filterEmptyAttrs(routerKey.attributes ?? {}) }
    }
    case "child": {
      return {
        child: {
          name: routerKey.name,
          attributes: filterEmptyAttrs(routerKey.attributes ?? {}),
        },
      }
    }
  }
}

const filterEmptyAttrs = (attrs: Attributes) =>
  Object.keys(attrs).reduce((acc, key) => {
    const val = attrs[key]
    if (val) {
      acc[key] = val
    }
    return acc
  }, {} as Attributes)
