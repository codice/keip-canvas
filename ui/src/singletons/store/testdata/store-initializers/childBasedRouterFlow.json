{
  "nodes": [
    {
      "id": "rn79OHQyAI",
      "type": "eipNode",
      "position": {
        "x": -23.429531055166507,
        "y": 14.944329221193257
      },
      "targetPosition": "left",
      "sourcePosition": "right",
      "data": {},
      "width": 128,
      "height": 132,
      "selected": false,
      "dragging": false
    },
    {
      "id": "cGUxaVOQ7L",
      "type": "eipNode",
      "position": {
        "x": 134.31436102080585,
        "y": 45.35279034427229
      },
      "targetPosition": "left",
      "sourcePosition": "right",
      "data": {},
      "width": 114,
      "height": 71,
      "selected": false,
      "dragging": false
    },
    {
      "id": "RJlDxzUi2a",
      "type": "eipNode",
      "position": {
        "x": 310.14028354278037,
        "y": 87.30725789068367
      },
      "targetPosition": "left",
      "sourcePosition": "right",
      "data": {},
      "width": 128,
      "height": 132,
      "selected": false,
      "dragging": false
    },
    {
      "id": "1su4DIeR80",
      "type": "eipNode",
      "position": {
        "x": 307.3643702586668,
        "y": -66.29327716359978
      },
      "targetPosition": "left",
      "sourcePosition": "right",
      "data": {},
      "width": 128,
      "height": 132,
      "selected": false,
      "dragging": false
    }
  ],
  "edges": [
    {
      "source": "rn79OHQyAI",
      "sourceHandle": "output",
      "target": "cGUxaVOQ7L",
      "targetHandle": "input",
      "id": "reactflow__edge-rn79OHQyAIoutput-cGUxaVOQ7Linput"
    },
    {
      "source": "cGUxaVOQ7L",
      "sourceHandle": "output",
      "target": "1su4DIeR80",
      "targetHandle": "input",
      "type": "dynamicEdge",
      "data": {
        "mapping": {
          "mapperName": "mapping",
          "matcher": {
            "name": "value",
            "type": "string",
            "description": "A value of the evaluation token that will be mapped to a channel reference (e.g., mapping value='foo' channel='myChannel')",
            "required": false
          },
          "matcherValue": "jms"
        }
      },
      "animated": true,
      "id": "reactflow__edge-cGUxaVOQ7Loutput-1su4DIeR80input",
      "selected": false
    },
    {
      "source": "cGUxaVOQ7L",
      "sourceHandle": "output",
      "target": "RJlDxzUi2a",
      "targetHandle": "input",
      "type": "dynamicEdge",
      "data": {
        "mapping": {
          "mapperName": "mapping",
          "matcher": {
            "name": "value",
            "type": "string",
            "description": "A value of the evaluation token that will be mapped to a channel reference (e.g., mapping value='foo' channel='myChannel')",
            "required": false
          },
          "matcherValue": "kafka"
        }
      },
      "animated": true,
      "id": "reactflow__edge-cGUxaVOQ7Loutput-RJlDxzUi2ainput",
      "selected": false
    }
  ],
  "eipConfigs": {
    "rn79OHQyAI": {
      "attributes": {},
      "children": [],
      "eipId": {
        "namespace": "ftp",
        "name": "inbound-channel-adapter"
      }
    },
    "cGUxaVOQ7L": {
      "attributes": {},
      "children": [],
      "eipId": {
        "namespace": "xml",
        "name": "xpath-router"
      },
      "routerKey": {
        "name": "xpath-expression",
        "attributes": {
          "expression": "/target/protocol"
        }
      }
    },
    "RJlDxzUi2a": {
      "attributes": {},
      "children": [],
      "eipId": {
        "namespace": "kafka",
        "name": "outbound-channel-adapter"
      }
    },
    "1su4DIeR80": {
      "attributes": {},
      "children": [],
      "eipId": {
        "namespace": "jms",
        "name": "outbound-channel-adapter"
      }
    }
  },
  "selectedChildNode": null,
  "layout": {
    "orientation": "horizontal",
    "density": "comfortable"
  }
}
