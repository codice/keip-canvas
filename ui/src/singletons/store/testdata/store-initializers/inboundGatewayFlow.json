{
  "nodes": [
    {
      "id": "_k0Z45PtEc",
      "type": "eipNode",
      "position": { "x": 0, "y": 0 },
      "data": { "followerId": "hF-dVMmHqZ", "label": "testIn" },
      "measured": { "width": 109, "height": 116 },
      "selected": false,
      "dragging": false,
      "targetPosition": "left",
      "sourcePosition": "right"
    },
    {
      "id": "hF-dVMmHqZ",
      "type": "followerNode",
      "position": { "x": 358, "y": 12 },
      "data": { "leaderId": "_k0Z45PtEc" },
      "measured": { "width": 114, "height": 92 },
      "selected": false,
      "dragging": false,
      "targetPosition": "left",
      "sourcePosition": "right"
    },
    {
      "id": "QvmTf6Jm4O",
      "type": "eipNode",
      "position": { "x": 184, "y": 13 },
      "data": {},
      "measured": { "width": 99, "height": 90 },
      "selected": false,
      "dragging": false,
      "targetPosition": "left",
      "sourcePosition": "right"
    }
  ],
  "edges": [
    {
      "source": "_k0Z45PtEc",
      "sourceHandle": "output",
      "target": "QvmTf6Jm4O",
      "targetHandle": "input",
      "id": "xy-edge___k0Z45PtEcoutput-QvmTf6Jm4Oinput"
    },
    {
      "source": "QvmTf6Jm4O",
      "sourceHandle": "output",
      "target": "hF-dVMmHqZ",
      "targetHandle": "input",
      "id": "xy-edge__QvmTf6Jm4Ooutput-hF-dVMmHqZinput"
    }
  ],
  "eipConfigs": {
    "_k0Z45PtEc": {
      "attributes": {},
      "children": [],
      "eipId": { "namespace": "http", "name": "inbound-gateway" }
    },
    "hF-dVMmHqZ": {
      "attributes": {},
      "children": [],
      "eipId": { "namespace": "integration", "name": "channel" }
    },
    "QvmTf6Jm4O": {
      "attributes": {},
      "children": [],
      "eipId": { "namespace": "integration", "name": "transformer" }
    }
  },
  "version": "1.0"
}
