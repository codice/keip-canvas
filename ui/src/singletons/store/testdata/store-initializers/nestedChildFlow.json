{
  "nodes": [
    {
      "id": "FL5Tssm8tV",
      "type": "eipNode",
      "position": {
        "x": 240.4526764062681,
        "y": 178.7952780472253
      },
      "targetPosition": "left",
      "sourcePosition": "right",
      "data": {},
      "width": 109,
      "height": 121,
      "selected": true,
      "dragging": false
    }
  ],
  "edges": [],
  "eipConfigs": {
    "FL5Tssm8tV": {
      "attributes": {},
      "children": ["43FAdk5SdBR", "hSZG9DyYy8q", "DD7ZbYQ3AST"],
      "eipId": {
        "namespace": "file",
        "name": "outbound-gateway"
      }
    },
    "43FAdk5SdBR": {
      "eipId": {
        "namespace": "integration",
        "name": "poller"
      },
      "attributes": {},
      "children": ["9OymVfY4n4p"]
    },
    "hSZG9DyYy8q": {
      "eipId": {
        "namespace": "integration",
        "name": "transactional"
      },
      "attributes": {},
      "children": []
    },
    "DD7ZbYQ3AST": {
      "eipId": {
        "namespace": "integration",
        "name": "request-handler-advice-chain"
      },
      "attributes": {},
      "children": ["5x4O_wYO-tM", "LYutVnb04dY"]
    },
    "9OymVfY4n4p": {
      "eipId": {
        "namespace": "integration",
        "name": "advice-chain"
      },
      "attributes": {},
      "children": ["gCL0YkrFGgW"]
    },
    "gCL0YkrFGgW": {
      "eipId": {
        "namespace": "integration",
        "name": "ref"
      },
      "attributes": {},
      "children": []
    },
    "5x4O_wYO-tM": {
      "eipId": {
        "namespace": "integration",
        "name": "retry-advice"
      },
      "attributes": {},
      "children": []
    },
    "LYutVnb04dY": {
      "eipId": {
        "namespace": "integration",
        "name": "ref"
      },
      "attributes": {},
      "children": []
    }
  },
  "selectedChildNode": null,
  "layout": {
    "orientation": "horizontal",
    "density": "comfortable"
  }
}
