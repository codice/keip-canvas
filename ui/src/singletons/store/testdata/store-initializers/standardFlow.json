{
  "nodes": [
    {
      "id": "9KWCqlIyy7",
      "type": "eipNode",
      "position": {
        "x": 0,
        "y": 91.25
      },
      "targetPosition": "left",
      "sourcePosition": "right",
      "data": {
        "label": "inbound"
      },
      "width": 114,
      "height": 129,
      "selected": false,
      "dragging": false
    },
    {
      "id": "LoiC2CFbLP",
      "type": "eipNode",
      "position": {
        "x": 201.02071435884494,
        "y": 121.16003882408089
      },
      "targetPosition": "left",
      "sourcePosition": "right",
      "data": {
        "label": "test-router"
      },
      "width": 114,
      "height": 71,
      "selected": false,
      "dragging": false
    },
    {
      "id": "H4-ED4F0XT",
      "type": "eipNode",
      "position": {
        "x": 402.5684310475592,
        "y": -19.39612977438884
      },
      "targetPosition": "left",
      "sourcePosition": "right",
      "data": {
        "label": "httpOut"
      },
      "width": 128,
      "height": 132,
      "selected": false,
      "dragging": false
    },
    {
      "id": "SV43RVeijQ",
      "type": "eipNode",
      "position": {
        "x": 378.5346234085371,
        "y": 244.49918423048507
      },
      "targetPosition": "left",
      "sourcePosition": "right",
      "data": {
        "label": "test-filter"
      },
      "width": 114,
      "height": 77,
      "selected": false,
      "dragging": false
    },
    {
      "id": "MrMneIdthg",
      "type": "eipNode",
      "position": {
        "x": 557.741165016793,
        "y": 220.00829459816538
      },
      "targetPosition": "left",
      "sourcePosition": "right",
      "data": {
        "label": "fileOut"
      },
      "width": 128,
      "height": 140,
      "selected": false,
      "dragging": false
    },
    {
      "id": "MZ1rIWIK3s",
      "type": "eipNode",
      "position": {
        "x": 526.8005098062063,
        "y": 400.71416837852064
      },
      "targetPosition": "left",
      "sourcePosition": "right",
      "data": {},
      "width": 111,
      "height": 121,
      "selected": false,
      "dragging": false
    }
  ],
  "edges": [
    {
      "source": "9KWCqlIyy7",
      "sourceHandle": "output",
      "target": "LoiC2CFbLP",
      "targetHandle": "input",
      "id": "reactflow__edge-9KWCqlIyy7output-LoiC2CFbLPinput"
    },
    {
      "source": "LoiC2CFbLP",
      "sourceHandle": "output",
      "target": "H4-ED4F0XT",
      "targetHandle": "input",
      "type": "dynamicEdge",
      "data": {
        "mapping": {
          "mapperId": {
            "namespace": "integration",
            "name": "mapping"
          },
          "matcher": {
            "name": "value",
            "type": "string",
            "description": "A value of the evaluation token that will be mapped to a channel reference (e.g., mapping value='foo' channel='myChannel')",
            "required": false
          },
          "isDefaultMapping": true
        }
      },
      "animated": true,
      "id": "reactflow__edge-LoiC2CFbLPoutput-H4-ED4F0XTinput",
      "selected": false
    },
    {
      "source": "LoiC2CFbLP",
      "sourceHandle": "output",
      "target": "SV43RVeijQ",
      "targetHandle": "input",
      "type": "dynamicEdge",
      "data": {
        "mapping": {
          "mapperId": {
            "namespace": "integration",
            "name": "mapping"
          },
          "matcher": {
            "name": "value",
            "type": "string",
            "description": "A value of the evaluation token that will be mapped to a channel reference (e.g., mapping value='foo' channel='myChannel')",
            "required": false
          },
          "matcherValue": "file"
        }
      },
      "animated": true,
      "id": "reactflow__edge-LoiC2CFbLPoutput-SV43RVeijQinput",
      "selected": false
    },
    {
      "source": "SV43RVeijQ",
      "sourceHandle": "output",
      "target": "MrMneIdthg",
      "targetHandle": "input",
      "id": "reactflow__edge-SV43RVeijQoutput-MrMneIdthginput"
    },
    {
      "source": "SV43RVeijQ",
      "sourceHandle": "discard",
      "target": "MZ1rIWIK3s",
      "targetHandle": "input",
      "id": "reactflow__edge-SV43RVeijQdiscard-MZ1rIWIK3sinput"
    }
  ],
  "eipConfigs": {
    "9KWCqlIyy7": {
      "attributes": {},
      "children": ["mcyTryMPewJ"],
      "eipId": {
        "namespace": "integration",
        "name": "inbound-channel-adapter"
      },
      "description": "message incoming"
    },
    "LoiC2CFbLP": {
      "attributes": {
        "phase": "init"
      },
      "children": [],
      "eipId": {
        "namespace": "integration",
        "name": "router"
      },
      "routerKey": {
        "eipId": {
          "namespace": "integration",
          "name": "expression"
        },
        "attributes": {
          "expression": "headers.protocol"
        }
      }
    },
    "H4-ED4F0XT": {
      "attributes": {},
      "children": [],
      "eipId": {
        "namespace": "http",
        "name": "outbound-channel-adapter"
      },
      "description": "send an http message"
    },
    "SV43RVeijQ": {
      "attributes": {
        "expression": "headers.filename != null"
      },
      "children": [],
      "eipId": {
        "namespace": "integration",
        "name": "filter"
      }
    },
    "MrMneIdthg": {
      "attributes": {
        "directory": "testdir"
      },
      "children": ["V1ls9ri4szs"],
      "eipId": {
        "namespace": "file",
        "name": "outbound-channel-adapter"
      },
      "description": "write to file"
    },
    "MZ1rIWIK3s": {
      "attributes": {
        "level": "WARN"
      },
      "children": [],
      "eipId": {
        "namespace": "integration",
        "name": "logging-channel-adapter"
      }
    },
    "mcyTryMPewJ": {
      "eipId": {
        "name": "poller",
        "namespace": "integration"
      },
      "attributes": {
        "fixed-rate": "2000"
      },
      "children": []
    },
    "V1ls9ri4szs": {
      "eipId": {
        "name": "transactional",
        "namespace": "file"
      },
      "attributes": {},
      "children": []
    }
  },
  "selectedChildNode": null,
  "layout": {
    "orientation": "horizontal",
    "density": "comfortable"
  },
  "customEntities": {}
}
