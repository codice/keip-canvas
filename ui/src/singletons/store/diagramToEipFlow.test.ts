import { act, renderHook } from "@testing-library/react"
import { expect, test, vi } from "vitest"
import { useEipFlow } from "./diagramToEipFlow"
import { resetMockStore } from "./storeTestingUtils"
import childBasedRouterFlow from "./testdata/store-initializers/childBasedRouterFlow.json"
import inboundGatewayFlow from "./testdata/store-initializers/inboundGatewayFlow.json"
import nestedChildFlow from "./testdata/store-initializers/nestedChildFlow.json"
import standardFlow from "./testdata/store-initializers/standardFlow.json"

vi.mock("zustand")

test("standard diagram to EipFlow success", () => {
  act(() => {
    resetMockStore(standardFlow)
  })

  const { result } = renderHook(() => useEipFlow())
  expect(result.current).toMatchSnapshot()
})

test("diagram includes a router with child based matcher to EipFlow success", () => {
  act(() => {
    resetMockStore(childBasedRouterFlow)
  })

  const { result } = renderHook(() => useEipFlow())
  expect(result.current).toMatchSnapshot()
})

test("diagram with deep child nesting", () => {
  act(() => {
    resetMockStore(nestedChildFlow)
  })

  const { result } = renderHook(() => useEipFlow())
  expect(result.current).toMatchSnapshot()
})

test("diagram with inbound-request-reply node", () => {
  act(() => {
    resetMockStore(inboundGatewayFlow)
  })

  const { result } = renderHook(() => useEipFlow())
  expect(result.current).toMatchSnapshot()
})
