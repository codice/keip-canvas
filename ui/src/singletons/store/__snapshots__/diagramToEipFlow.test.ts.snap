// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`diagram includes a router with child based matcher to EipFlow success 1`] = `
{
  "edges": [
    {
      "id": "ch-rn79OHQyAI-cGUxaVOQ7L",
      "source": "rn79OHQyAI",
      "target": "cGUxaVOQ7L",
      "type": "default",
    },
    {
      "id": "ch-cGUxaVOQ7L-1su4DIeR80",
      "source": "cGUxaVOQ7L",
      "target": "1su4DIeR80",
      "type": "default",
    },
    {
      "id": "ch-cGUxaVOQ7L-RJlDxzUi2a",
      "source": "cGUxaVOQ7L",
      "target": "RJlDxzUi2a",
      "type": "default",
    },
  ],
  "nodes": [
    {
      "attributes": {},
      "children": [],
      "connectionType": "source",
      "description": undefined,
      "eipId": {
        "name": "inbound-channel-adapter",
        "namespace": "ftp",
      },
      "id": "rn79OHQyAI",
      "role": "endpoint",
    },
    {
      "attributes": {},
      "children": [
        {
          "attributes": {
            "expression": "/target/protocol",
          },
          "name": "xpath-expression",
        },
        {
          "attributes": {
            "channel": "ch-cGUxaVOQ7L-1su4DIeR80",
            "value": "jms",
          },
          "name": "mapping",
        },
        {
          "attributes": {
            "channel": "ch-cGUxaVOQ7L-RJlDxzUi2a",
            "value": "kafka",
          },
          "name": "mapping",
        },
      ],
      "connectionType": "content_based_router",
      "description": undefined,
      "eipId": {
        "name": "xpath-router",
        "namespace": "int-xml",
      },
      "id": "cGUxaVOQ7L",
      "role": "router",
    },
    {
      "attributes": {},
      "children": [],
      "connectionType": "sink",
      "description": undefined,
      "eipId": {
        "name": "outbound-channel-adapter",
        "namespace": "kafka",
      },
      "id": "RJlDxzUi2a",
      "role": "endpoint",
    },
    {
      "attributes": {},
      "children": [],
      "connectionType": "sink",
      "description": undefined,
      "eipId": {
        "name": "outbound-channel-adapter",
        "namespace": "jms",
      },
      "id": "1su4DIeR80",
      "role": "endpoint",
    },
  ],
}
`;

exports[`diagram with deep child nesting 1`] = `
{
  "edges": [],
  "nodes": [
    {
      "attributes": {},
      "children": [
        {
          "attributes": {},
          "children": [
            {
              "attributes": {},
              "children": [
                {
                  "attributes": {},
                  "name": "ref",
                },
              ],
              "name": "advice-chain",
            },
          ],
          "name": "poller",
        },
        {
          "attributes": {},
          "name": "transactional",
        },
        {
          "attributes": {},
          "children": [
            {
              "attributes": {},
              "name": "retry-advice",
            },
            {
              "attributes": {},
              "name": "ref",
            },
          ],
          "name": "request-handler-advice-chain",
        },
      ],
      "connectionType": "request_reply",
      "description": undefined,
      "eipId": {
        "name": "outbound-gateway",
        "namespace": "file",
      },
      "id": "FL5Tssm8tV",
      "role": "endpoint",
    },
  ],
}
`;

exports[`diagram with inbound-request-reply node 1`] = `
{
  "edges": [
    {
      "id": "ch-testIn-QvmTf6Jm4O",
      "source": "testIn",
      "target": "QvmTf6Jm4O",
      "type": "default",
    },
    {
      "id": "ch-QvmTf6Jm4O-testIn-reply-channel",
      "source": "QvmTf6Jm4O",
      "target": "testIn-reply-channel",
      "type": "default",
    },
    {
      "id": "ch-testIn-reply-channel-testIn",
      "source": "testIn-reply-channel",
      "target": "testIn",
      "type": "default",
    },
  ],
  "nodes": [
    {
      "attributes": {},
      "children": [],
      "connectionType": "inbound_request_reply",
      "description": undefined,
      "eipId": {
        "name": "inbound-gateway",
        "namespace": "http",
      },
      "id": "testIn",
      "role": "endpoint",
    },
    {
      "attributes": {},
      "children": [],
      "connectionType": "passthru",
      "description": undefined,
      "eipId": {
        "name": "channel",
        "namespace": "integration",
      },
      "id": "testIn-reply-channel",
      "role": "channel",
    },
    {
      "attributes": {},
      "children": [],
      "connectionType": "passthru",
      "description": undefined,
      "eipId": {
        "name": "transformer",
        "namespace": "integration",
      },
      "id": "QvmTf6Jm4O",
      "role": "transformer",
    },
  ],
}
`;

exports[`standard diagram to EipFlow success 1`] = `
{
  "edges": [
    {
      "id": "ch-inbound-test-router",
      "source": "inbound",
      "target": "test-router",
      "type": "default",
    },
    {
      "id": "ch-test-router-httpOut",
      "source": "test-router",
      "target": "httpOut",
      "type": "default",
    },
    {
      "id": "ch-test-router-test-filter",
      "source": "test-router",
      "target": "test-filter",
      "type": "default",
    },
    {
      "id": "ch-test-filter-fileOut",
      "source": "test-filter",
      "target": "fileOut",
      "type": "default",
    },
    {
      "id": "ch-test-filter-MZ1rIWIK3s",
      "source": "test-filter",
      "target": "MZ1rIWIK3s",
      "type": "discard",
    },
  ],
  "nodes": [
    {
      "attributes": {},
      "children": [
        {
          "attributes": {
            "fixed-rate": "2000",
          },
          "name": "poller",
        },
      ],
      "connectionType": "source",
      "description": "message incoming",
      "eipId": {
        "name": "inbound-channel-adapter",
        "namespace": "integration",
      },
      "id": "inbound",
      "role": "endpoint",
    },
    {
      "attributes": {
        "default-output-channel": "ch-test-router-httpOut",
        "expression": "headers.protocol",
        "phase": "init",
      },
      "children": [
        {
          "attributes": {
            "channel": "ch-test-router-test-filter",
            "value": "file",
          },
          "name": "mapping",
        },
      ],
      "connectionType": "content_based_router",
      "description": undefined,
      "eipId": {
        "name": "router",
        "namespace": "integration",
      },
      "id": "test-router",
      "role": "router",
    },
    {
      "attributes": {},
      "children": [],
      "connectionType": "sink",
      "description": "send an http message",
      "eipId": {
        "name": "outbound-channel-adapter",
        "namespace": "http",
      },
      "id": "httpOut",
      "role": "endpoint",
    },
    {
      "attributes": {
        "expression": "headers.filename != null",
      },
      "children": [],
      "connectionType": "tee",
      "description": undefined,
      "eipId": {
        "name": "filter",
        "namespace": "integration",
      },
      "id": "test-filter",
      "role": "router",
    },
    {
      "attributes": {
        "directory": "testdir",
      },
      "children": [
        {
          "attributes": {},
          "name": "transactional",
        },
      ],
      "connectionType": "sink",
      "description": "write to file",
      "eipId": {
        "name": "outbound-channel-adapter",
        "namespace": "file",
      },
      "id": "fileOut",
      "role": "endpoint",
    },
    {
      "attributes": {
        "level": "WARN",
      },
      "children": [],
      "connectionType": "sink",
      "description": undefined,
      "eipId": {
        "name": "logging-channel-adapter",
        "namespace": "integration",
      },
      "id": "MZ1rIWIK3s",
      "role": "endpoint",
    },
  ],
}
`;
